### PR TITLE
rgw: by default allow dynamic resharding to go to 4999 shards

### PR DESCRIPTION
**Pull request by ivancich**
Given that we now allow dynamic resharding to reduce the number of shards and the general stability of the dynamic resharding functionality, we should give allow around 5000 shards, which in turn help support buckets of around 500,000,000 entries.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ivancich** [comment]
I did some studies to see how important the prime aspect of shard count is. I took the texts of Oliver Twist and A Tale of Two Cities from Project Gutenberg, filtered out lines that were fewer than 20 characters, and ran the lines through sha256sum, and then did a modulo of a prime and an adjacent non-prime (e.g., 101 and 100, 307 and 308, 1001 and 1000). I then did a standard deviation on the modulo values, which would be the shard numbers. The standard deviations were very similar between prime and non-prime, which I interpreted as they were distributing equally well.

A Tale of Two Cities
* 307 6.54932
* 308 6.06735

Oliver Twist
* 307 6.71842
* 308 7.05512

We use xxhash rather than sha256 in the code, so I should probably try that hash as well. But if the that two distributes equally well, then perhaps we can remove the prime preference.

**Comment by cbodley** [comment]
@ivancich it's hard for me to evaluate this 4999 number objectively, but my general concern is just that bucket listing performance degrades with the shard count

we've seen extreme cases in the past where clusters hit a performance "cliff" once the latency of a bucket listing request exceeds the client's timeout value. once clients start retrying these requests, it further increases the latency of each request and results in even more timeouts/retries. at this point, the bucket effectively becomes unlistable and breaks client applications

it's really hard to say exactly where this "cliff" is because it depends on hardware and load. but i tend to view `rgw_max_dynamic_shards=1999` as a conservative value that should prevent that from happening on most clusters

but thinking more on listing performance, the main scaling problem here seems to be `rgw_bucket_index_max_aio=128` which limits how many shard-listing rados requests we'll issue at a time. we raised that default from 8 to 128 in https://github.com/ceph/ceph/pull/28558, but that still seems tiny compared to 4999 shards. 4999/128 means _at least_ 39 osd round trips of total latency per ordered listing

in contrast, the rados cluster itself has available parallelism up to the pg count of its bucket index pool, which could be in the thousands. so what if we gave bucket listing unlimited concurrency? then we'd only be limited by the Objecter's inflight ops throttle. the more we can reduce the latency of bucket listing, the larger buckets we can reasonably support

to summarize, i would love to find a default configuration for rgw that can scale to 1 billion objects and beyond. but i feel like we need some data to inform this decision. can we make a plan to measure performance of writes and listings at various shard counts to see how it scales? ideally we would do this on a huge/fast cluster and push the performance as far as a we can, then again on a small cluster to see what default configuration is reasonable for most/all users

**Comment by cbodley** [comment]
p.s. https://bugzilla.redhat.com/show_bug.cgi?id=2290684 recently pointed out that multisite's 8x multiplier doesn't behave sanely at the high end, so we're relying on the default limit of 1999 to keep things relatively sane there. i think we've struggled to reliably replicate buckets even at 1999 shards, so i don't think we should go any higher in multisite

**Comment by cbodley** [comment]
in https://github.com/ceph/ceph/pull/59222 i propose raising the default value of `rgw_bucket_index_max_aio` to ~4k. but we can test various values in the meantime without needing that pr

**Comment by cbodley** [comment]
bump?

**Comment by markhpc** [comment]
@ivancich Should reopen this one?
